addons/web_editor/static/src/js/common/utils.js

```diff
@@ -375,6 +375,32 @@ function _addBackgroundImageAttributes(...newAttributes) {
 function _isBackgroundImageAttribute(attribute) {
     return BACKGROUND_IMAGE_ATTRIBUTES.has(attribute);
 }
+/**
+ * Checks if an element supposedly marked with the o_editable_media class should
+ * in fact be editable (checks if its environment looks like a non editable
+ * environment whose media should be editable).
+ *
+ * TODO: the name of this function is voluntarily bad to reflect the fact that
+ * this system should be improved. The combination of o_not_editable,
+ * o_editable, getContentEditableAreas, getReadOnlyAreas and other concepts
+ * related to what should be editable or not should be reviewed.
+ *
+ * @returns {boolean}
+ */
+function _shouldEditableMediaBeEditable(mediaEl) {
+    // Some sections of the DOM are contenteditable="false" (for
+    // example with the help of the o_not_editable class) but have
+    // inner media that should be editable (the fact the container
+    // is not is to prevent adding text in between those medias).
+    // This case is complex and the solution to support it is not
+    // perfect: we mark those media with a class and check that the
+    // first non editable ancestor is in fact in an editable parent.
+    const parentEl = mediaEl.parentElement;
+    const nonEditableAncestorRootEl = parentEl && parentEl.closest('[contenteditable="false"]');
+    return nonEditableAncestorRootEl
+        && nonEditableAncestorRootEl.parentElement
+        && nonEditableAncestorRootEl.parentElement.isContentEditable;
```

Comment:
> This is wrong IMO and, if I understand correctly, in that case this function would return false instead of true, right ?

As advertised throughout the comments and the commit, the system is not perfect and cannot handle everything. But the case you mention there... should be working, and should already return false, why do you say it does not? I don't understand.

> I don't understand why you want to double check the o_editable_media. Why not trust it ? If the media has the magic class then you can change it, regardless of the context in which you are, thus removing the need for this function entirely. What are the odds we'll end up with this class on an image that is located in a truly non-savable context ?

Don't see the reason for coding with "odds" in mind for this case. The class is added on the snippets directly and is not removed on save etc. So this simple case does not seem to be an oddly case to me:
- Have a "description" html field on some record
- Somehow somewhere, drag and drop a snippet with an o_editable_media inside
- On another page somehow somewhere, this field content is displayed (thanks to a t-out or whatever) in a non editable area => I cannot magically remove the o_editable_media classes in there.

But even without this simple "use of a field" case, I feel this would be necessary: people will copy/paste snippet DOM here and there, we should not expect them to reliably modify those technical stuff inside.
That's actually also my point: we should have a better system to reliably decide what is editable or not, without the need of having to be careful on how we structure the snippets and where we put them. For this "team" snippet, the notion of "the picture are editable but their container is not" should be easy to describe and the DOM to reflect that should be the same wherever the snippet is put. In this case, I thought of replacing this o_editable_media system by a o_not_editable_container_expect_medias class, but in the end it solved some things at the cost of others. The whole system should be improved, relying on DOM classes hierarchy is not enough anymore. If not in pure JS method overrides, I could very well see a definition of snippet being made in such a way it describes what area are editable or not (kinda like o_not_editable / o_editable_media does) but not with classes dropped and saved in the DOM.

> If you really want to double check that nonetheless, why not do mediaEl.closest('[contenteditable="true"]') ? 

Exactly for the reason of the example that you say does not work above... but that does work thanks to my function :grimacing:

Reply:
Ok then I don't understand the case. If I understand correctly, you are saying that it should return `false` in this case:
```html
<div contentEditable="true">
    <div contentEditable="false">
        <div contentEditable="false">
            <img class="o_editable_media"/>
        </div>
    </div>
</div>
```
but `true` in this one:
```html
<div contentEditable="true">
    <div contentEditable="false">
        <img class="o_editable_media"/>
    </div>
</div>
```

Why ? Why would the fact that there are two contentEditable="false" level of parents have an effect on `o_editable_media` ?

Reply:
Heuristic according to the real cases, and which again would require a better system to describe them.

- Having a root with `contenteditable="true"` (not `contentEditable` by the way :wink:) is the base case for the majority of the DOM: everything is editable by default in Odoo
- The final `contenteditable="false"` (in both your case) combined with the editable-marked media is what we are trying to describe: a non editable container meant to hold an editable media
- Any `contenteditable="false"` just above should probably mean the whole section is not editable, media or not. Again because someone creating a non editable section (who knows, with dynamic stuff etc) can t-out snippets with "editable media".

Your confusion is about the o_editable_media notion. The name should be o_editable_media_despite_we_dont_want_text_around_it. Again, same as my idea to not have this but rather a o_not_editable_except_existing_media_inside container. Any way is ugly, any way has limits, master should be improved.

Reply:
Ok I got the theory of it, thanks for the explanation.

Isn't there another way we can detect the intermediary non-editable section ? Like maybe it would have some branding due to the `t-out` or something ? If not then I guess your code is fine since, as you say, it's an heuristic and we probably don't yet have any actual use cases of `o_editable_media` in which this heuristic would not stand.

In fact, your idea of `o_not_editable_except_existing_media_inside` seems way easier to understand so now I'm kinda sad we're not implementing that instead... ^^' Though to be fair it would have the same problem if someone copies a div with this class and it has some media in it, you'd still argue to bulletproof the checks in case the div happens to be in an environment where the media should still not be editable despite the class so... I guess it's not really _that_ better. At least not enough to continue delaying this fix IMO.

Reply:
> I guess it's not really that better.

Indeed like I said, better for some things, worse for others. After starting implementing it, I went back to o_editable_media so we don't say I changed the original code for no reason :wink: Again, IMO, all of this should disappear. Once we have an OWL editor in master, some related tasks merged and stable cases working, I'll write some specs to improve all of this.

> Like maybe it would have some branding due to the t-out or something ?

Yes, the simple t-out case should work on its own without this, kinda by chance as there are multiple o_editable_media entry point in the code that all work by checking the editor root their own way (same as o_not_editable: there should not be that many occurences of it in JS logic: another proof that things have to be better refactored later). But:
- t-out + t-ignore
- lazy loading field content in JS (there are cases in eLearning, we recently had to patch code around that too)
- ...

Reply:
It would be nice to have a complete spec of all the weird possible cases that can be encountered indeed. Let's go for this fix then, with the agreed upon changes on master and then I can't wait for a proper spec for a future refactor :p